docs/architecture/adr-049-state-sync-hooks.md

```diff
@@ -0,0 +1,170 @@
+# ADR 049: State Sync Hooks
+
+## Changelog
+
+- Jan 19, 2022: Initial Draft
+
+## Status
+
+Draft, Under Implementation
+
+## Abstract
+
+This ADR provides hooks for app modules to publish snapshot of additional state(outside of IAVL tree) for state-sync.
+
+## Context
+
+New clients use state-sync to download snapshots of module state from peer nodes. Currently, the snapshot consists of a
+stream of `SnapshotStoreItem` and `SnapshotIAVLItem`, which means for app modules that maintain their states outside of
+the IAVL tree, they can not add their states to the snapshot stream for state-sync.
+
+## Decision
+
+A simple proposal based on our existing implementation is that, we can add two new message types: `SnapshotExtensionMeta` 
+and `SnapshotExtensionPayload`, and they are appended to the existing multi-store stream with `SnapshotExtensionMeta` 
+acting as a delimiter between extensions.
+
+```proto
+// SnapshotItem is an item contained in a rootmulti.Store snapshot.
+message SnapshotItem {
+  // item is the specific type of snapshot item.
+  oneof item {
+    SnapshotStoreItem        store             = 1;
+    SnapshotIAVLItem         iavl              = 2 [(gogoproto.customname) = "IAVL"];
+    SnapshotExtensionMeta    extension         = 3;
+    SnapshotExtensionPayload extension_payload = 4;
+  }
+}
+
+// SnapshotStoreItem contains metadata about a snapshotted store.
+message SnapshotStoreItem {
+  string name = 1;
+}
+
+// SnapshotIAVLItem is an exported IAVL node.
+message SnapshotIAVLItem {
+  bytes key     = 1;
+  bytes value   = 2;
+  int64 version = 3;
+  int32 height  = 4;
+}
+
+// SnapshotExtensionMeta contains metadata about an external snapshotter.
+message SnapshotExtensionMeta {
+  string name   = 1;
+  uint32 format = 2;
```

Comment:
format is globally unique?
like x/wasm needs one, ethermint another, agoric a 3rd?
Or is this used within one namespace?

Reply:
It’s used within the snapshotter/namespace.

Reply:
They don't need to globally unique. They need to be unique in the state machine. So if you are 3 different modules that use different types of state sync packets you need formats 1, 2, 3. I don't see why they should be global

Reply:
yes, they are not global, they can be 1 at the same time.

Reply:
@zmanian and @yihuang you are saying different things.

> Zaki: They need to be unique in the state machine

This was more or less what I meant by globally unique (on the chain). That you cannot have 2 modules on the same chain both using format 1.

> Yi Huang: yes, they are not global, they can be 1 at the same time.

This is what I understood the answer to be, that `(name, format)` must be unique within this state machine. (And if we keep name globally unique, then we only have to consider format in the context of our own module).

Reply:
Maybe you two can discuss this and determine which interpretation is correct (I am inclined to accept Yi Huang's answer as he is working on this code, but Zaki knows details about the state sync abci framework that the rest of us are missing)

Reply:
To clarify there are two formats at play in the snapshot:
- one is the snapshot format currently used, it decides both the serialization/chunking and rootmulti snapshot formats.
- then it's the extension snapshotter formats, they are only used by each extension snapshotter to decide the format of payloads included in `SnapshotExtensionPayload` message. So they are not related to one another.

Reply:
> This is what I understood the answer to be, that (name, format) must be unique within this state machine. (And if we keep name globally unique, then we only have to consider format in the context of our own module).

technically speaking we can have duplicated `name`s in the stream, it just means the snapshotter's `Restore` methods will be called multiple times. But I don't think that makes much sense, maybe we should even validate to prevent that situation.

Reply:
This is somewhat confusing. I recommend updating the field comment to be very explicit.

Reply:
Up. It would be good to clarify if `name` is unique, or `(name, format)` is. IMO first one is simpler.

Reply:
> Up. It would be good to clarify if `name` is unique, or `(name, format)` is. IMO first one is simpler.

`name` is unique in current implementation.